src/types.d.ts

```diff
@@ -0,0 +1,161 @@
+declare type EventName =
+  | 'initialized'
+  | 'context.initialized'
+  | 'context.changed'
+  | 'context.value.removed'
+  | 'context.value.added'
+  | 'context.value.changed'
+  | 'context.destroyed'
+  | 'genome.request.sent'
+  | 'config.request.sent'
+  | 'genome.request.received'
+  | 'config.request.received'
+  | 'request.failed'
+  | 'genome.updated'
+  | 'config.updated'
+  | 'effective.genome.updated'
+  | 'store.destroyed'
+  | 'confirmed'
+  | 'contaminated'
+  | 'event.emitted';
+
+declare type Listener = (event: string, ...args: any[]) => void;
+
+export class MiniPromise<T = any> extends Promise<T> {
+  static createPromise<T>(executor: (resolve: (value: T | PromiseLike<T>) => void, reject: (reason?: any) => void) => void): MiniPromise<T>;
+}
+
+declare interface SubscribablePromise<T = any> extends Promise<T> {
+  listen(handler: (value: T) => void): () => void;
+}
+
+export interface EvolvClientOptions {
+  environment: string;
+  endpoint?: string;
+  lazyUid?: boolean;
+  requireConsent?: boolean;
+  js?: boolean;
+  css?: boolean;
+  pushstate?: boolean;
+  version?: number;
+  autoConfirm?: boolean;
+  analytics?: boolean;
+  bufferEvents?: boolean;
+  clientName?: string;
+  uid?: string;
+  sid?: string;
+  useCookies?: boolean;
+  timeout?: number;
+  hooks?: {};
+}
+
+export class EvolvClient {
+  context: EvolvContext;
+  environment: string;
+
+  constructor(options: Partial<EvolvClientOptions>);
+
+  preload(prefixed: string[], configOnly?: boolean, immediate?: boolean): void;
+
+  get<T = any>(key: string): SubscribablePromise<T | Error>;
+
+  on(key: EventName, listener: Listener): void;
+
+  once(topic: EventName, listener: Listener): void;
+
+  initialize(uid: string, remoteContext?: Record<string, any>, localContext?: Record<string, any>): void;
+
+  confirm(): void;
+
+  contaminate(details?: Record<string, any>, allExperiments?: boolean): void;
+
+  emit(type: string, metadata?: Record<string, any>, flush?: boolean): void;
+
+  isActive(key: string): SubscribablePromise<boolean>;
+
+  getActiveKeys(prefix?: string): SubscribablePromise<{ current: string[]; previous: string[] }>;
+
+  /** @deprecated */
+  clearActiveKeys(prefix?: string): void;
+
+  reevaluateContext(): void;
+
+  getConfig(key: string): SubscribablePromise<any>;
+
+  flush(): void;
+
+  allowEvents(): void;
+
+  destroy(): void;
+}
+
+interface Allocation {
+  uid: string;
+  eid: string;
+  cid: string;
+  ordinal: number;
+  group_id: string;
+  excluded: boolean;
+}
+
+interface Confirmation {
+  cid: string;
+  timestamp: number;
+}
+
+export interface RemoteContext extends Record<string, any> {
+  confirmations: Confirmation[];
+  keys: { active: string };
+  variants: { active: string[] };
+  experiments: {
+    allocations: Allocation[];
+    confirmations: Confirmation[];
+    exclusions: any[];
+  };
+  device: string;
+  location: string;
+  platform: string;
+  geo: {
```

Comment:
Should these be part of the interface? They are dependent on the info we send back from configuration. Do we explicitly set them in the javascript-sdk codebase?

Reply:
Hmm... Explicitly no. They are spliced in from _configuration.json_.

Reply:
I split out the context from PAPI into a separate interface that `RemoteContext` extends from. I have another PR inbound that depends on this, so I'm going to merge this one. I can remove the interface in that PR if you still feel uneasy about it.